File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/021c0aca45fd5ea8d2817d965cb74f2c.py
from sys import stdin, stdout
from math import sqrt

def mp(x, y):
    return (x, y)

def sz(v):
    return len(v)

def raz(a, b, M):
    c = a - b
    return c + M if c < 0 else c

def sum(a, b, M):
    c = a + b
    return c - M if c >= M else c

def id17(r):
    return 32 - (r.bit_length() - 1)

def euc(m, n):
    if n == 0:
        return (1 if m >= 0 else -1, 0)
    t = m // n
    ans1 = euc(n, m - t * n)
    return (ans1[1], ans1[0] - ans1[1] * t)

def prod(a, b, M):
    return (a * b) % M

def id1(p1, p2, n):
    if p1[0] == -1 or p2[0] == -1:
        return (-1, -1)
    np1 = (0, raz(p1[1], p1[0], n))
    np2 = (raz(p2[0], p1[0], n), raz(p2[1], p1[0], n))
    if np2[0] <= np2[1]:
        if np2[1] == n - 1:
            return (-1, -1)
        else:
            return (sum(np1[0], p1[0], n), sum(max(np1[1], np2[1]), p1[0], n))
    else:
        if np2[0] <= np1[1] + 1:
            return (-1, -1)
        else:
            return (sum(np2[0], p1[0], n), sum(max(np1[1], np2[1]), p1[0], n))

class Sparse:
    def __init__(self):
        self.n = 0
        self.sp = []

    def construct(self, v):
        self.n = sz(v)
        r = id17(self.n - 1)
        self.sp = [[(0, 0)] * self.n for _ in range(r)]
        for i in range(self.n):
            self.sp[0][i] = v[i]
        for i in range(1, r):
            for j in range(self.n):
                p1 = self.sp[i - 1][j]
                p2 = self.sp[i - 1][(j + (1 << (i - 1))) % self.n]
                self.sp[i][j] = id1(p1, p2, self.n)

    def segment(self, l, r):
        if l == -1 or r == -1:
            return (-1, -1)
        szz = raz(r, l, self.n)
        if l == r:
            return self.sp[0][l]
        g = id17(szz) - 1
        return id1(self.sp[g][l], self.sp[g][(r + 1 - (1 << g)) % self.n], self.n)

def main():
    input = stdin.read
    data = input().split()
    n = int(data[0])
    if n == 1:
        stdout.write("0\n")
        return

    r = list(map(int, data[1:n + 1]))
    zz = id17(n) + 1
    v = [Sparse() for _ in range(zz)]
    prs = [(0, 0)] * n

    for i in range(n):
        if r[i] >= n - 1:
            prs[i] = (-1, -1)
        else:
            prs[i] = id1((raz(i, r[i], n), i), (i, sum(i, r[i], n)), n)

    v[0].construct(prs)

    for i in range(1, zz):
        id14 = [(0, 0)] * n
        for j in range(n):
            pairr = v[i - 1].segment(j, j)
            id14[j] = v[i - 1].segment(pairr[0], pairr[1])
        v[i].construct(id14)

    result = []
    for i in range(n):
        my_pr = (i, i)
        ans = 0
        for j in range(zz - 1, -1, -1):
            new_pr = v[j].segment(my_pr[0], my_pr[1])
            if new_pr[0] != -1:
                ans += (1 << j)
                my_pr = new_pr
        result.append(ans + 1)

    stdout.write(' '.join(map(str, result)) + '\n')

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_4000/C++/C/021c0aca45fd5ea8d2817d965cb74f2c.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

#define MAX_N 100000
#define MAX_LOG 32

typedef struct {
    int X, Y;
} pii;

typedef struct {
    pii sp[MAX_LOG][MAX_N];
    int n;
} sparse;

int raz(int a, int b, int M) {
    int c = a - b;
    return c < 0 ? c + M : c;
}

int sum(int a, int b, int M) {
    int c = a + b;
    return c >= M ? c - M : c;
}

pii unite_seg(pii p1, pii p2, int n) {
    if ((p1.X == -1) || (p2.X == -1))
        return (pii){-1, -1};
    pii np1 = {0, raz(p1.Y, p1.X, n)};
    pii np2 = {raz(p2.X, p1.X, n), raz(p2.Y, p1.X, n)};
    if (np2.X <= np2.Y) {
        if (np2.Y == n - 1)
            return (pii){-1, -1};
        else
            return (pii){sum(np1.X, p1.X, n), sum(max(np1.Y, np2.Y), p1.X, n)};
    } else {
        if (np2.X <= np1.Y + 1)
            return (pii){-1, -1};
        else
            return (pii){sum(np2.X, p1.X, n), sum(max(np1.Y, np2.Y), p1.X, n)};
    }
}

void sparse_construct(sparse *s, pii *v, int n) {
    s->n = n;
    int r = 0;
    while ((1 << r) < n) r++;
    for (int i = 0; i < n; i++)
        s->sp[0][i] = v[i];
    for (int i = 1; i <= r; i++) {
        for (int j = 0; j < n; j++) {
            pii p1 = s->sp[i - 1][j];
            pii p2 = s->sp[i - 1][(j + (1 << (i - 1))) % n];
            s->sp[i][j] = unite_seg(p1, p2, n);
        }
    }
}

pii sparse_segment(sparse *s, int l, int r, int n) {
    if ((l == -1) || (r == -1))
        return (pii){-1, -1};
    int szz = raz(r, l, n);
    if (l == r)
        return s->sp[0][l];
    int g = 0;
    while ((1 << g) <= szz) g++;
    g--;
    return unite_seg(s->sp[g][l], s->sp[g][((r + 1 - (1 << g)) % n + n) % n], n);
}

int count_tro(int r) {
    return 32 - __builtin_clz(r);
}

int main() {
    int n;
    scanf("%d", &n);
    if (n == 1) {
        printf("0\n");
        return 0;
    }
    int r[MAX_N];
    for (int i = 0; i < n; i++)
        scanf("%d", &r[i]);
    
    int zz = count_tro(n) + 1;
    sparse v[MAX_LOG];
    pii prs[MAX_N];

    for (int i = 0; i < n; i++) {
        if (r[i] >= n - 1)
            prs[i] = (pii){-1, -1};
        else
            prs[i] = unite_seg((pii){raz(i, r[i], n), i}, (pii){i, sum(i, r[i], n)}, n);
    }

    sparse_construct(&v[0], prs, n);

    for (int i = 1; i < zz; i++) {
        pii new_prs[MAX_N];
        for (int j = 0; j < n; j++) {
            pii pairr = sparse_segment(&v[i - 1], j, j, n);
            new_prs[j] = sparse_segment(&v[i - 1], pairr.X, pairr.Y, n);
        }
        sparse_construct(&v[i], new_prs, n);
    }

    for (int i = 0; i < n; i++) {
        pii my_pr = {i, i};
        int ans = 0;
        for (int j = zz - 1; j >= 0; --j) {
            pii new_pr = sparse_segment(&v[j], my_pr.X, my_pr.Y, n);
            if (new_pr.X != -1) {
                ans += (1 << j);
                my_pr = new_pr;
            }
        }
        printf("%d ", ans + 1);
    }
    printf("\n");
    return 0;
}
